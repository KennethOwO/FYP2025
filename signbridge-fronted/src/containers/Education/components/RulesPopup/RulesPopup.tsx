import React, { useEffect, useRef } from 'react';
import styles from './RulesPopup.module.css';
import buttonClickedSound from '/music/btnClicked.wav';

interface RulesPopupProps {
  onClose: () => void;
  title: string;
  rules: string[];
}

const RulesPopup: React.FC<RulesPopupProps> = ({ onClose, title, rules }) => {
  const popupRef = useRef<HTMLDivElement>(null);

  // Function to play button clicked sound
  const playButtonClickedSound = () => {
    const audio = new Audio(buttonClickedSound);
    audio.play();
  };

  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (popupRef.current && !popupRef.current.contains(event.target as Node)) {
        onClose(); 
      }
    };

    document.addEventListener('mousedown', handleClickOutside);
    return () => {
      document.removeEventListener('mousedown', handleClickOutside);
    };
  }, [onClose]);

  return (
    <div className={styles.popup_container}>
      <div ref={popupRef} className={styles.popup_content}>
        <h3>{title}</h3>
        <button className={styles.close_btn_pushable} role="button" onClick={() => {
          onClose();
          playButtonClickedSound(); 
        }}>
          <span className={styles.close_btn_shadow}></span>
          <span className={styles.close_btn_edge}></span>
          <span className={styles.close_btn_front}>
            <i className={`${styles.fa_close_btn} fa fa-close`}></i>
          </span>
        </button>
        
        <div className={styles.popup_details}>
          {rules.map((rule, index) => (
            <p key={index}>&#8226; {rule}</p>
          ))}
        </div>
      </div>
    </div>
  );
};

export default RulesPopup;
