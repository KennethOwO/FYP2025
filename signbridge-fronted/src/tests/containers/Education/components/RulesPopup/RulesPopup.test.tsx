import RulesPopup from "@root/containers/Education/components/RulesPopup/RulesPopup";
import { create } from "react-test-renderer";
import { cleanup } from "@testing-library/react";

describe("Test RulesPopup", () => {
  afterEach(() => {
    // Clean up the DOM
    cleanup();
    // Reset mocks to their initial state
    jest.resetAllMocks();
  });
  it("should render correctly", () => {
    const props = {
      onClose: jest.fn(),
      title: "This is the Rules Popup",
      rules: ["Rule 1", "Rule 2", "Rule 3"],
    };
    const tree = create(<RulesPopup {...props} />);
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
