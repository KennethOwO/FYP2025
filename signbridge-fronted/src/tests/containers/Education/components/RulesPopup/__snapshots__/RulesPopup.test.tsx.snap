// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test RulesPopup should render correctly 1`] = `
<div
  className="popup_container"
>
  <div
    className="popup_content"
  >
    <h3>
      This is the Rules Popup
    </h3>
    <button
      className="close_btn_pushable"
      onClick={[Function]}
      role="button"
    >
      <span
        className="close_btn_shadow"
      />
      <span
        className="close_btn_edge"
      />
      <span
        className="close_btn_front text"
      >
        <i
          className="fa fa-close fa-close-btn"
        />
      </span>
    </button>
    <div
      className="popup_details"
    >
      <p>
        • 
        Rule 1
      </p>
      <p>
        • 
        Rule 2
      </p>
      <p>
        • 
        Rule 3
      </p>
    </div>
  </div>
</div>
`;
